src/memory/RomAscii16X.cc

```diff
@@ -0,0 +1,135 @@
+// ASCII-X 16kB cartridges
+//
+// Banks:
+//  first  16kB: 0x4000 - 0x7FFF and 0xC000 - 0xFFFF
+//  second 16kB: 0x8000 - 0xBFFF and 0x0000 - 0x3FFF
+//
+// Address to change banks:
+//  first  16kB: 0x6000 - 0x6FFF, 0xA000 - 0xAFFF, 0xE000 - 0xEFFF, 0x2000 - 0x2FFF
+//  second 16kB: 0x7000 - 0x7FFF, 0xB000 - 0xBFFF, 0xF000 - 0xFFFF, 0x3000 - 0x3FFF
+//
+// 12-bit bank number is composed by address bits 8-11 and the data bits.
```

Comment:
Not a comment on this PR, but a comment on the mapper design.

Using 4 address bits to "extend" the data-bus to 12-bits is a bit unusual. And unexpected (to me). From a hardware point of view I understand how this could indeed lead to a simpler design. From a software point of view I have mixed feelings:
* Switching to a fixed bank is only two Z80 instructions: `ld a,<n> ; ld (<nn>),a` (or equivalent).
* But switching to a dynamic bank (e.g. a calculated bank number for a large lookup table) is more complex. Maybe something like this:
```
Input: 12-bit bank number in HL
    ld a,h
    or 0x60   ; 0x70 for 2nd bank
    ld h,a
    ld (hl),l
```
So actually not too bad, and maybe that `or 0x60` could be pre-calculated.

Is it a hard requirement that the switch regions are compatible with the ASCII16 mapper? I often like it that the switch region for a bank is located in the bank itself. That allows to work with the mapper when it's only partially selected in the MSX slots.

Anyway, just my first impression from reading this documentation.

Reply:
> Using 4 address bits to "extend" the data-bus to 12-bits is a bit unusual. And unexpected (to me). From a hardware point of view I understand how this could indeed lead to a simpler design. From a software point of view I have mixed feelings: [...] `ld (hl),l` [...] So actually not too bad, and maybe that or 0x60 could be pre-calculated.

Yes, indeed if you want to do a 16-bit bank switch you’d do `ld (hl),l` (that’s why it uses address bits 8-11), which is only 7 cycles for a 16-bit switch. You’ll typically want to embed the register address in the bank number.

Another approach (which Usas2 will take probably) is to divide the game data into two 4 MB sections, e.g. level data in the lower half and music data in the upper half. Then continue to use 8-bit bank numbers within them and you’d write to `0x6000` and `0x6100` depending on the type of data.

From a hardware perspective I receive 24 bits (address + data) in a single operation, so I can use them all in one go rather than having two operations. Two operations means they need to be stored in separate registers. This would require two additional 74LVC823 chips (relatively expensive) as well as an extra decoder chip.

So the mapper’s design involves both hardware considerations in terms of cost (minimal chip count, 2-layer board, and components on a single side), as well as software considerations (being fast and versatile).

> Is it a hard requirement that the switch regions are compatible with the ASCII16 mapper?

Not per se, I started out with ASCII16 because it is the largest capacity mapper, also fairly simple, and I thought it would add some value if it were compatible. Both in terms of making it easier to use or adapt existing developer tooling and code, as well as to make it useful as a platform to run or release ASCII16 games on.

Usas2 code currectly built for ASCII16 runs on it verbatim, so does Aleste, Penguin Wars, Pointless Fighting, etc.

Also note that ASCII16 is flash-able because you can write to memory without affecting its bank register. The Konami mapper bank registers overlap with the memory they switch, so they can’t be flashed normally.

> I often like it that the switch region for a bank is located in the bank itself. That allows to work with the mapper when it's only partially selected in the MSX slots.

The bank registers are mirrored, so you can access them from any page.

The only downside of this mirroring is that they’re also mirrored to page 3, which means the BIOS’s expanded slot detection messes up the initial bank of page 2. Unfortunately not something I can solve without adding extra chips or removing the mirroring entirely. It bugs me from a theoretical point of view that it’s no longer _fully_ ASCII16 compatible, but in practice there’s no real problem, even for existing ASCII16 games that I tried, so I kept it as-is.